src/uffd_handler/src/handler.rs

```diff
@@ -65,18 +67,32 @@ impl UffdManager for Uffd {
     fn populate_from_file(
         &self,
         buff: *const libc::c_void,
+        start_addr: usize,
+        end_addr: usize,
         region: &MemRegion,
     ) -> Result<(usize, usize)> {
-        let src = buff as u64 + region.mapping.offset;
-        let start_addr = region.mapping.base_host_virt_addr;
-        let len = region.mapping.size;
-        // Populate whole region from backing mem-file.
-        // This offers an example of how memory can be loaded in RAM,
-        // however this can be adjusted to accommodate use case needs.
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+        // The result will always be positive because both start_addr and end_addr are within the
+        // current region and the difference between them is always at least page size.
+        let mut len = end_addr - start_addr;
+        // Length to uffd copy must be multiple of page size.
+        let reminder = len % page_size;
+        if reminder != 0 {
+            len += page_size - reminder;
+        }
+
+        // Compute source to prefault the memory chunk from.
+        let prefault_offset = start_addr - region.mapping.base_host_virt_addr;
+        let src = buff as u64 + region.mapping.offset + prefault_offset as u64;
+
         // SAFETY: this is safe because the parameters are valid.
         let ret = unsafe {
-            self.copy(src as *const _, start_addr as *mut _, len, true)
-                .map_err(HandlerError::UffdCopy)?
+            match self.copy(src as *const _, start_addr as *mut _, len, true) {
```

Comment:
nit: not sure if it's idiomatic in Rust, but do we actually need to wrap the pattern matching in `unsafe` too? I guess it is only `copy()` that is unsafe here?

Reply:
That is true, there is no need to wrap the entire `match` into `unsafe`. Fixed.